src/types.h

```diff
@@ -438,41 +445,41 @@ inline int from_to(Move m) {
  return m & 0xFFF;
 }
 
-inline bool is_gating(Move m) {
-  return (m & (3 << 12)) && (rank_of(from_sq(m)) == RANK_1 || rank_of(from_sq(m)) == RANK_8) && ((m & (3 << 14)) == NORMAL || (m & (3 << 14)) == CASTLING || rank_of(from_sq(m)) == rank_of(to_sq(m)));
+inline MoveType type_of(Move m) {
+  return MoveType(m & CASTLING ? (m ^ (m >> 6)) & PROMOTION : m & ENPASSANT);
```

Comment:
This is a place where a flag like `CASTLING_OR_PROMOTION` might help to avoid confusion.

Reply:
I agree. The problem with this is that CASTLING_OR_PROMOTION would be visible throughout the entire source but would only have meaning within types.h. I could do as follows:

```
#define CASTLING_OR_PROMOTION (1 << 12)
#define FLAGS (3 << 12)
#define GATE_ON_TO (1 << 13)
```
and then undef them at the end of the file. Do you think this is a good approach?